File: ts/sb/struct_sim_order.ts
import * as rt from "./type"
import * as _ from "./_"

export interface SimOrder extends rt.Serializable, rt.Deserializable {
    id: number;
    accountId: number;
    itemId: number;
    // 办理人姓名
    name: string;
    // 联系电话
    phone: string;
    // 身份证号
    idNo: string;
    // 所在城市
    cityCode: number;
    // 详细地址
    address: string;
    // 新手机号码
    newPhone: string;
    // 佣金
    commission: number;
    status: _.OrderStatus;
}

const simOrderHeaderWidths = [1, 1, 1, 2, 2, 2, 1, 2, 2, 1, 1] as const;

export const newSimOrder = (): SimOrder => {
    const s = {
        id: 0,
        accountId: 0,
        itemId: 0,
        name: "",
        phone: "",
        idNo: "",
        cityCode: 0,
        address: "",
        newPhone: "",
        commission: 0,
        status: _.DefaultOrderStatus(),
    } as any as SimOrder;
    s.set = (buf: rt.Buffer) => setSimOrder(buf, s);
    s.get = (buf: rt.Buffer) => {
        const [next, err] = getSimOrder(buf);
        if (err === undefined) Object.assign(s, next);
        return err;
    };
    return s;
};

export const isZeroSimOrder = (s: SimOrder | null | undefined): boolean => {
    if (s === null || s === undefined) return true;
    if (s.id !== 0) return false;
    if (s.accountId !== 0) return false;
    if (s.itemId !== 0) return false;
    if (s.name !== "") return false;
    if (s.phone !== "") return false;
    if (s.idNo !== "") return false;
    if (s.cityCode !== 0) return false;
    if (s.address !== "") return false;
    if (s.newPhone !== "") return false;
    if (s.commission !== 0) return false;
    if (!_.IsDefaultOrderStatus(s.status as any)) return false;
    return true;
};

export const validateSimOrder = (s: SimOrder | null | undefined): rt.Err => {
    if (s === null || s === undefined) return undefined;
    { const [, err] = rt.textState(s.name); if (err !== null) return new Error(`Name: ${err.message}`); }
    { const [, err] = rt.textState(s.phone); if (err !== null) return new Error(`Phone: ${err.message}`); }
    { const [, err] = rt.textState(s.idNo); if (err !== null) return new Error(`IdNo: ${err.message}`); }
    { const [, err] = rt.textState(s.address); if (err !== null) return new Error(`Address: ${err.message}`); }
    { const [, err] = rt.textState(s.newPhone); if (err !== null) return new Error(`NewPhone: ${err.message}`); }
    if (!_.IsAssignableOrderStatus(s.status as any)) return new Error(`Status 非法枚举值: ${s.status as any}`);
    return undefined;
};

export const getSimOrder = (buf: rt.Buffer): [SimOrder, rt.Err] => {
    const s = newSimOrder();
    const headerBits = 16;
    const [header, errHeader] = buf.read(rt.headerSize(headerBits));
    if (errHeader !== null) return [s, new Error(`not enough data`)];
    const [headerStates, errHeaderState] = rt.readHeader(header, simOrderHeaderWidths, "SimOrder header");
    if (errHeaderState !== undefined) return [s, errHeaderState];
    const idPresent = headerStates[0] === 1;
    const accountIdPresent = headerStates[1] === 1;
    const itemIdPresent = headerStates[2] === 1;
    const nameState = headerStates[3];
    const phoneState = headerStates[4];
    const idNoState = headerStates[5];
    const cityCodePresent = headerStates[6] === 1;
    const addressState = headerStates[7];
    const newPhoneState = headerStates[8];
    const commissionPresent = headerStates[9] === 1;
    const statusPresent = headerStates[10] === 1;
    if (idPresent) {
        const [value, err] = rt.getU32(buf);
        if (err !== null) return [s, new Error(`get SimOrder Id: ${err.message}`)];
        s.id = value as any;
    }
    if (accountIdPresent) {
        const [value, err] = rt.getU32(buf);
        if (err !== null) return [s, new Error(`get SimOrder AccountId: ${err.message}`)];
        s.accountId = value as any;
    }
    if (itemIdPresent) {
        const [value, err] = rt.getU32(buf);
        if (err !== null) return [s, new Error(`get SimOrder ItemId: ${err.message}`)];
        s.itemId = value as any;
    }
    { const [value, err] = rt.getTextCompact(buf, nameState); if (err !== null) return [s, new Error(`get SimOrder Name: ${err.message}`)]; s.name = value; }
    { const [value, err] = rt.getTextCompact(buf, phoneState); if (err !== null) return [s, new Error(`get SimOrder Phone: ${err.message}`)]; s.phone = value; }
    { const [value, err] = rt.getTextCompact(buf, idNoState); if (err !== null) return [s, new Error(`get SimOrder IdNo: ${err.message}`)]; s.idNo = value; }
    if (cityCodePresent) {
        const [value, err] = rt.getU32(buf);
        if (err !== null) return [s, new Error(`get SimOrder CityCode: ${err.message}`)];
        s.cityCode = value as any;
    }
    { const [value, err] = rt.getTextCompact(buf, addressState); if (err !== null) return [s, new Error(`get SimOrder Address: ${err.message}`)]; s.address = value; }
    { const [value, err] = rt.getTextCompact(buf, newPhoneState); if (err !== null) return [s, new Error(`get SimOrder NewPhone: ${err.message}`)]; s.newPhone = value; }
    if (commissionPresent) {
        const [value, err] = rt.getU16(buf);
        if (err !== null) return [s, new Error(`get SimOrder Commission: ${err.message}`)];
        s.commission = value as any;
    }
    s.status = _.DefaultOrderStatus();
    if (statusPresent) {
        const [value, err] = rt.getU8(buf);
        if (err !== null) return [s, new Error(`get SimOrder Status: ${err.message}`)];
        const item = value as _.OrderStatus;
        if (!_.IsOrderStatus(item)) return [s, new Error(`get SimOrder Status: 非法枚举值: ${item}`)];
        s.status = item;
    }
    const errValidate = validateSimOrder(s);
    if (errValidate !== undefined) return [s, new Error(`validate failed: ${errValidate.message}`)];
    return [s, undefined];
};

export const setSimOrder = (buf: rt.Buffer, s: SimOrder): rt.Err => {
    if (s === null || s === undefined) return new Error(`set SimOrder: value is null or undefined`);
    const errValidate = validateSimOrder(s);
    if (errValidate !== undefined) return new Error(`validate SimOrder: ${errValidate.message}`);
    const startOffset = buf.writeOffset;
    const [nameState, errNameState] = rt.textState(s.name);
    if (errNameState !== null) return errNameState;
    const [phoneState, errPhoneState] = rt.textState(s.phone);
    if (errPhoneState !== null) return errPhoneState;
    const [idNoState, errIdNoState] = rt.textState(s.idNo);
    if (errIdNoState !== null) return errIdNoState;
    const [addressState, errAddressState] = rt.textState(s.address);
    if (errAddressState !== null) return errAddressState;
    const [newPhoneState, errNewPhoneState] = rt.textState(s.newPhone);
    if (errNewPhoneState !== null) return errNewPhoneState;
    const headerStates = [
        s.id !== 0 ? 1 : 0,
        s.accountId !== 0 ? 1 : 0,
        s.itemId !== 0 ? 1 : 0,
        nameState,
        phoneState,
        idNoState,
        s.cityCode !== 0 ? 1 : 0,
        addressState,
        newPhoneState,
        s.commission !== 0 ? 1 : 0,
        !_.IsDefaultOrderStatus(s.status as any) ? 1 : 0,
    ];
    const [header, errHeader] = rt.writeHeader(simOrderHeaderWidths, headerStates);
    if (errHeader !== undefined) { buf.rewindWrite(startOffset); return new Error(`set header: ${errHeader.message}`); }
    const errHeaderWrite = buf.write(header);
    if (errHeaderWrite !== null) { buf.rewindWrite(startOffset); return errHeaderWrite; }
    if (s.id !== 0) {
        const err = rt.setU32(buf, s.id as any);
        if (err !== null) { buf.rewindWrite(startOffset); return new Error(`set SimOrder Id: ${err.message}`); }
    }
    if (s.accountId !== 0) {
        const err = rt.setU32(buf, s.accountId as any);
        if (err !== null) { buf.rewindWrite(startOffset); return new Error(`set SimOrder AccountId: ${err.message}`); }
    }
    if (s.itemId !== 0) {
        const err = rt.setU32(buf, s.itemId as any);
        if (err !== null) { buf.rewindWrite(startOffset); return new Error(`set SimOrder ItemId: ${err.message}`); }
    }
    { const err = rt.setTextCompact(buf, nameState, s.name); if (err !== null) { buf.rewindWrite(startOffset); return new Error(`set SimOrder Name: ${err.message}`); } }
    { const err = rt.setTextCompact(buf, phoneState, s.phone); if (err !== null) { buf.rewindWrite(startOffset); return new Error(`set SimOrder Phone: ${err.message}`); } }
    { const err = rt.setTextCompact(buf, idNoState, s.idNo); if (err !== null) { buf.rewindWrite(startOffset); return new Error(`set SimOrder IdNo: ${err.message}`); } }
    if (s.cityCode !== 0) {
        const err = rt.setU32(buf, s.cityCode as any);
        if (err !== null) { buf.rewindWrite(startOffset); return new Error(`set SimOrder CityCode: ${err.message}`); }
    }
    { const err = rt.setTextCompact(buf, addressState, s.address); if (err !== null) { buf.rewindWrite(startOffset); return new Error(`set SimOrder Address: ${err.message}`); } }
    { const err = rt.setTextCompact(buf, newPhoneState, s.newPhone); if (err !== null) { buf.rewindWrite(startOffset); return new Error(`set SimOrder NewPhone: ${err.message}`); } }
    if (s.commission !== 0) {
        const err = rt.setU16(buf, s.commission as any);
        if (err !== null) { buf.rewindWrite(startOffset); return new Error(`set SimOrder Commission: ${err.message}`); }
    }
    if (!_.IsDefaultOrderStatus(s.status as any)) {
        const err = rt.setU8(buf, _.NormalizeOrderStatus(s.status as any) as any);
        if (err !== null) { buf.rewindWrite(startOffset); return new Error(`set SimOrder Status: ${err.message}`); }
    }
    return undefined;
};

export const readSimOrder = (buf: rt.Buffer): [SimOrder, rt.Err] => getSimOrder(buf);

export const eqSimOrder = (a: SimOrder | null | undefined, b: SimOrder | null | undefined): boolean => {
    if (isZeroSimOrder(a as any) && isZeroSimOrder(b as any)) return true;
    if (a === null || a === undefined || b === null || b === undefined) return false;
    if (!rt.eqU32(a.id, b.id)) return false;
    if (!rt.eqU32(a.accountId, b.accountId)) return false;
    if (!rt.eqU32(a.itemId, b.itemId)) return false;
    if (!rt.eqText(a.name, b.name)) return false;
    if (!rt.eqText(a.phone, b.phone)) return false;
    if (!rt.eqText(a.idNo, b.idNo)) return false;
    if (!rt.eqU32(a.cityCode, b.cityCode)) return false;
    if (!rt.eqText(a.address, b.address)) return false;
    if (!rt.eqText(a.newPhone, b.newPhone)) return false;
    if (!rt.eqU16(a.commission, b.commission)) return false;
    if (!_.eqOrderStatusValue(a.status as any, b.status as any)) return false;
    return true;
};

export const getSimOrderListBody = (buf: rt.Buffer, state: number): [SimOrder[], rt.Err] => {
    const [list, err] = rt.getBitmapListCompact<SimOrder>(
        buf,
        state,
        () => newSimOrder(),
        (buf) => readSimOrder(buf),
    );
    return [list, err];
};

export const setSimOrderListBody = (buf: rt.Buffer, state: number, v: SimOrder[]): rt.Err => {
    return rt.setBitmapListCompact<SimOrder>(
        buf,
        state,
        v,
        (item) => isZeroSimOrder(item),
        (buf, item) => setSimOrder(buf, item),
    );
}

File: go/sb/struct_sim_order.go
package sb

import (
	"bytes"
	"fmt"
	"slices"
)

type SimOrder struct {
	Id uint32 `bson:"id" json:"id"`
	AccountId uint32 `bson:"account_id" json:"account_id"`
	ItemId uint32 `bson:"item_id" json:"item_id"`
	// 办理人姓名
	Name string `bson:"name" json:"name"`
	// 联系电话
	Phone string `bson:"phone" json:"phone"`
	// 身份证号
	IdNo string `bson:"id_no" json:"id_no"`
	// 所在城市
	CityCode uint32 `bson:"city_code" json:"city_code"`
	// 详细地址
	Address string `bson:"address" json:"address"`
	// 新手机号码
	NewPhone string `bson:"new_phone" json:"new_phone"`
	// 佣金
	Commission uint16 `bson:"commission" json:"commission"`
	Status OrderStatus `bson:"status" json:"status"`
}

func sizeSimOrderBody(s *SimOrder, bits []byte) (int, error) {
	if s == nil { return 0, fmt.Errorf("sizeSimOrder: nil value") }
	bodySize := 0
	if s.Id != 0 {
		bodySize += 4
		if bits != nil { SetBit(bits, uint8(0), true) }
	}
	if s.AccountId != 0 {
		bodySize += 4
		if bits != nil { SetBit(bits, uint8(1), true) }
	}
	if s.ItemId != 0 {
		bodySize += 4
		if bits != nil { SetBit(bits, uint8(2), true) }
	}
	if s.Name != "" {
		fieldSize, err := sizeText(s.Name)
		if err != nil { return 0, fmt.Errorf("sizeSimOrder Name: %w", err) }
		bodySize += fieldSize
		if bits != nil { SetBit(bits, uint8(3), true) }
	}
	if s.Phone != "" {
		fieldSize, err := sizeText(s.Phone)
		if err != nil { return 0, fmt.Errorf("sizeSimOrder Phone: %w", err) }
		bodySize += fieldSize
		if bits != nil { SetBit(bits, uint8(4), true) }
	}
	if s.IdNo != "" {
		fieldSize, err := sizeText(s.IdNo)
		if err != nil { return 0, fmt.Errorf("sizeSimOrder IdNo: %w", err) }
		bodySize += fieldSize
		if bits != nil { SetBit(bits, uint8(5), true) }
	}
	if s.CityCode != 0 {
		bodySize += 4
		if bits != nil { SetBit(bits, uint8(6), true) }
	}
	if s.Address != "" {
		fieldSize, err := sizeText(s.Address)
		if err != nil { return 0, fmt.Errorf("sizeSimOrder Address: %w", err) }
		bodySize += fieldSize
		if bits != nil { SetBit(bits, uint8(7), true) }
	}
	if s.NewPhone != "" {
		fieldSize, err := sizeText(s.NewPhone)
		if err != nil { return 0, fmt.Errorf("sizeSimOrder NewPhone: %w", err) }
		bodySize += fieldSize
		if bits != nil { SetBit(bits, uint8(8), true) }
	}
	if s.Commission != 0 {
		bodySize += 2
		if bits != nil { SetBit(bits, uint8(9), true) }
	}
	if s.Status != 0 {
		bodySize += 1
		if bits != nil { SetBit(bits, uint8(10), true) }
	}
	return bodySize, nil
}

func sizeSimOrder(s *SimOrder) (int, error) {
	bodySize, err := sizeSimOrderBody(s, nil)
	if err != nil { return 0, err }
	return 2 + bodySize, nil
}

func sizeSimOrderList(v SimOrderList) (int, error) {
	if len(v) > 65535 { return 0, fmt.Errorf("list length exceeds uint16 max") }
	totalSize := 2
	for i, item := range v {
		itemSize, err := sizeSimOrder(item)
		if err != nil { return 0, fmt.Errorf("sizeSimOrderList[%d]: %w", i, err) }
		totalSize += itemSize
	}
	return totalSize, nil
}

func GetSimOrder(buf *bytes.Buffer, s *SimOrder) error {
	if s == nil { return nil }
	const bitSize = 2
	if buf.Len() < bitSize { return fmt.Errorf("GetSimOrder bitmask: %d - %d", buf.Len(), bitSize) }
	bits := buf.Next(bitSize)
	if GetBit(bits, uint8(0)) {
		val, err := GetU32(buf)
		if err != nil { return fmt.Errorf("GetSimOrder Id: %w", err) }
		s.Id = val
	}
	if GetBit(bits, uint8(1)) {
		val, err := GetU32(buf)
		if err != nil { return fmt.Errorf("GetSimOrder AccountId: %w", err) }
		s.AccountId = val
	}
	if GetBit(bits, uint8(2)) {
		val, err := GetU32(buf)
		if err != nil { return fmt.Errorf("GetSimOrder ItemId: %w", err) }
		s.ItemId = val
	}
	if GetBit(bits, uint8(3)) {
		val, err := GetText(buf)
		if err != nil { return fmt.Errorf("GetSimOrder Name: %w", err) }
		s.Name = val
	}
	if GetBit(bits, uint8(4)) {
		val, err := GetText(buf)
		if err != nil { return fmt.Errorf("GetSimOrder Phone: %w", err) }
		s.Phone = val
	}
	if GetBit(bits, uint8(5)) {
		val, err := GetText(buf)
		if err != nil { return fmt.Errorf("GetSimOrder IdNo: %w", err) }
		s.IdNo = val
	}
	if GetBit(bits, uint8(6)) {
		val, err := GetU32(buf)
		if err != nil { return fmt.Errorf("GetSimOrder CityCode: %w", err) }
		s.CityCode = val
	}
	if GetBit(bits, uint8(7)) {
		val, err := GetText(buf)
		if err != nil { return fmt.Errorf("GetSimOrder Address: %w", err) }
		s.Address = val
	}
	if GetBit(bits, uint8(8)) {
		val, err := GetText(buf)
		if err != nil { return fmt.Errorf("GetSimOrder NewPhone: %w", err) }
		s.NewPhone = val
	}
	if GetBit(bits, uint8(9)) {
		val, err := GetU16(buf)
		if err != nil { return fmt.Errorf("GetSimOrder Commission: %w", err) }
		s.Commission = val
	}
	if GetBit(bits, uint8(10)) {
		val, err := GetOrderStatus(buf)
		if err != nil { return fmt.Errorf("GetSimOrder Status: %w", err) }
		s.Status = val
		if !IsOrderStatus(s.Status) { return fmt.Errorf("GetSimOrder Status: 非法枚举值: %d", s.Status) }
	}
	if err := ValidateSimOrder(s); err != nil { return fmt.Errorf("ValidateSimOrder: %w", err) }
	return nil
}

func SetSimOrder(buf *bytes.Buffer, s *SimOrder) error {
	if s == nil { return fmt.Errorf("SetSimOrder: nil value") }
	if err := ValidateSimOrder(s); err != nil { return fmt.Errorf("ValidateSimOrder: %w", err) }
	const bitSize = 2
	var bits [2]byte
	bodySize, err := sizeSimOrderBody(s, bits[:])
	if err != nil { return err }
	buf.Grow(bitSize + bodySize)
	if _, err := buf.Write(bits[:]); err != nil { return fmt.Errorf("SetSimOrder write bitmask: %w", err) }
	if s.Id != 0 {
		if err := SetU32(buf, s.Id); err != nil { return fmt.Errorf("SetSimOrder Id: %w", err) }
	}
	if s.AccountId != 0 {
		if err := SetU32(buf, s.AccountId); err != nil { return fmt.Errorf("SetSimOrder AccountId: %w", err) }
	}
	if s.ItemId != 0 {
		if err := SetU32(buf, s.ItemId); err != nil { return fmt.Errorf("SetSimOrder ItemId: %w", err) }
	}
	if s.Name != "" {
		if err := SetText(buf, s.Name); err != nil { return fmt.Errorf("SetSimOrder Name: %w", err) }
	}
	if s.Phone != "" {
		if err := SetText(buf, s.Phone); err != nil { return fmt.Errorf("SetSimOrder Phone: %w", err) }
	}
	if s.IdNo != "" {
		if err := SetText(buf, s.IdNo); err != nil { return fmt.Errorf("SetSimOrder IdNo: %w", err) }
	}
	if s.CityCode != 0 {
		if err := SetU32(buf, s.CityCode); err != nil { return fmt.Errorf("SetSimOrder CityCode: %w", err) }
	}
	if s.Address != "" {
		if err := SetText(buf, s.Address); err != nil { return fmt.Errorf("SetSimOrder Address: %w", err) }
	}
	if s.NewPhone != "" {
		if err := SetText(buf, s.NewPhone); err != nil { return fmt.Errorf("SetSimOrder NewPhone: %w", err) }
	}
	if s.Commission != 0 {
		if err := SetU16(buf, s.Commission); err != nil { return fmt.Errorf("SetSimOrder Commission: %w", err) }
	}
	if s.Status != 0 {
		if err := SetOrderStatus(buf, s.Status); err != nil { return fmt.Errorf("SetSimOrder Status: %w", err) }
	}
	return nil
}

func ValidateSimOrder(s *SimOrder) error {
	if s == nil { return nil }
	if s.Status != 0 && !IsOrderStatus(s.Status) { return fmt.Errorf("Status 非法枚举值: %d", s.Status) }
	return nil
}

func EqSimOrder(a, b *SimOrder) bool {
	if a == b { return true }
	if a == nil || b == nil { return false }
	if !EqU32(a.Id, b.Id) { return false }
	if !EqU32(a.AccountId, b.AccountId) { return false }
	if !EqU32(a.ItemId, b.ItemId) { return false }
	if !EqText(a.Name, b.Name) { return false }
	if !EqText(a.Phone, b.Phone) { return false }
	if !EqText(a.IdNo, b.IdNo) { return false }
	if !EqU32(a.CityCode, b.CityCode) { return false }
	if !EqText(a.Address, b.Address) { return false }
	if !EqText(a.NewPhone, b.NewPhone) { return false }
	if !EqU16(a.Commission, b.Commission) { return false }
	if !(a.Status == b.Status) { return false }
	return true
}

// Standalone functions
func ReadSimOrder(buf *bytes.Buffer) (*SimOrder, error) {
	s := new(SimOrder)
	return s, GetSimOrder(buf, s)
}

type SimOrderList []*SimOrder
func GetSimOrderList(buf *bytes.Buffer) (SimOrderList, error) {
	count, err := GetU16(buf)
	if err != nil { return nil, err }
	list := make([]*SimOrder, count)
	for i := range list {
		item, err := ReadSimOrder(buf)
		if err != nil { return nil, err }
		list[i] = item
	}
	return SimOrderList(list), nil
}
func SetSimOrderList(buf *bytes.Buffer, v SimOrderList) error {
	totalSize, err := sizeSimOrderList(v)
	if err != nil { return err }
	buf.Grow(totalSize)
	if err := SetU16(buf, uint16(len(v))); err != nil { return err }
	for _, item := range v {
		if err := SetSimOrder(buf, item); err != nil { return err }
	}
	return nil
}
func ValidateSimOrderList(v SimOrderList) error {
	for i, item := range v {
		if item == nil { return fmt.Errorf("SimOrderList[%d]: nil item", i) }
		if err := ValidateSimOrder(item); err != nil { return fmt.Errorf("SimOrderList[%d]: %w", i, err) }
	}
	return nil
}
func EqSimOrderList(a, b SimOrderList) bool { return slices.EqualFunc(a, b, EqSimOrder) }
